File: python/blob/worldbank.py
import requests

class DataFetcher:
    def __init__(self, country: str):
        self.country = country
        
    def fetchData(self, indicator: str):
        api_url = f"https://api.worldbank.org/v2/countries/{self.country}/indicators/{indicator}?per_page=100&format=json"
        response = requests.get(api_url)
        if not response.status_code == 200:
            raise Exception(f"Error fetching data: {response.status_code}")
      
        return response.json()  
    
    
class DataProcessor:
    def __init__(self):
        # do some complicated init here that results in
        country = "USA"
        
        # create an object based on the previously constructed data
        self.fetcher = DataFetcher(country)
        
    def getGdpPerCapita(self) -> float:
        data = self.fetcher.fetchData("NY.GDP.PCAP.KD")
        return data[1][0]['value']
        
        
        

File: javascript/blob/worldbank.ts
export class DataFetcher {
    constructor(public country: string) {
        this.country = country;
    }
  
    async fetchData(indicator: string): Promise<any> {
      const apiUrl = `https://api.worldbank.org/v2/countries/${this.country}/indicators/${indicator}?per_page=100&format=json`;
      try {
        const response = await fetch(apiUrl);
        if (!response.ok) {
          throw new Error(`Error fetching data: ${response.status}`);
        }
        return response.json();
      } catch (error) {
        throw error; // Re-throw the error for handling
      }
    }
  }
  
  export class DataProcessor {
    fetcher: DataFetcher;

    constructor() {
      // Complicated initialization can go here, resulting in country = "USA"
      let country = "USA";

      // create an object based on the previously constructed data
      // Uncomment: this.fetcher = new DataFetcher(country)
      // After start
      this.fetcher = this.createFetcher(country);
    }
  
    createFetcher(country: string): DataFetcher {
      return new DataFetcher(country);
    }
    // After end
  
    async getGdpPerCapita(): Promise<number> {
      const data = await this.fetcher.fetchData("NY.GDP.PCAP.KD");
      return data[1][0].value;
    }

    // After start
    setFetcher(fetcher: DataFetcher) {
        this.fetcher = fetcher
    }
    // After end
  }